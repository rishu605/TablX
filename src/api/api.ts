const getProjectDetailsFromApi = async () => {
    const apiUrl = "https://raw.githubusercontent.com/saaslabsco/frontend-assignment/refs/heads/master/frontend-assignment.json"
    try {
        const response = await fetch(apiUrl)
        if(!response) {
            throw new Error("Error getting data")
        }

        const data = await response.json()
        if(!data) {
            throw new Error("Error parsing JSON data")
        }
        return data
    } catch(err) {
        console.error("Failed to get project details", err)
        throw err
    }
}

export {
    getProjectDetailsFromApi
}