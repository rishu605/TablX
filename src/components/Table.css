.table-container {
    padding: 20px;
    max-width: 500px;
    margin: 0 auto;
}

.styled-table {
    width: 100%;
    border-collapse: collapse;
    margin: 25px 0;
    font-size: 18px;
    text-align: left;
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);
}

.styled-table thead tr {
    background-color: #009879;
    color: #ffffff;
    text-align: center;
}

.styled-table th,
.styled-table td {
    padding: 12px 15px;
}

.styled-table tbody {
    display: block;
    height: 250px;
    overflow: hidden;
}

.col-width-small {
    width: 15%;
}

.styled-table tbody tr {
    background-color: #242424;
    color: #ffffff;
    height: 50px;
    display: flex;
    text-align: center;
}

.styled-table tbody tr:nth-of-type(even) {
    background-color: #ffffff;
    color: #242424;
}

.styled-table tbody tr:last-of-type {
    border-bottom: 2px solid #009879;
}

.styled-table thead,
.styled-table tbody tr {
    display: table;
    width: 100%;
    table-layout: fixed;
}

.pagination {
    display: flex;
    justify-content: center;
    margin-top: 20px;
}

.pagination button {
    background-color: #242424;
    border: 1px solid #ddd;
    padding: 10px 15px;
    margin: 0 5px;
    cursor: pointer;
}

.pagination button.active {
    background-color: #009879;
    color: #ffffff;
    border: 1px solid #009879;
}