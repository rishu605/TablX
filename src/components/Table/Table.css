.table-container {
    padding: 20px;
    max-width: 500px;
    margin: 0 auto;
}

.styled-table {
    width: 100%;
    border-collapse: collapse;
    margin: 25px 0;
    font-size: 18px;
    text-align: left;
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);
}

.styled-table thead tr {
    background-color: #009879;
    color: #ffffff;
    text-align: center;
}

.styled-table th,
.styled-table td {
    padding: 12px 15px;
}

.styled-table tbody {
    display: block;
    height: 250px;
    overflow: hidden;
}

.col-width-small {
    width: 15%;
}

.styled-table tbody tr {
    background-color: #242424;
    color: #ffffff;
    height: 50px;
    display: flex;
    text-align: center;
}

.styled-table tbody tr:nth-of-type(even) {
    background-color: #ffffff;
    color: #242424;
}

.styled-table tbody tr:last-of-type {
    border-bottom: 2px solid #009879;
}

.styled-table thead,
.styled-table tbody tr {
    display: table;
    width: 100%;
    table-layout: fixed;
}

.pagination {
    display: flex;
    justify-content: center;
    margin-top: 20px;
}

.pagination button {
    background-color: #242424;
    border: 1px solid #ddd;
    padding: 10px 15px;
    margin: 0 5px;
    cursor: pointer;
}

.pagination button.active {
    background-color: #009879;
    color: #ffffff;
    border: 1px solid #009879;
}
.error-message {
    color: #ffffff;
    background-color: #e74c3c; /* Soft red for error indication */
    border: 1px solid #c0392b; /* Darker red border for subtle emphasis */
    border-radius: 5px;
    padding: 12px;
    margin-top: 20px;
    text-align: center;
    font-size: 16px;
    font-weight: bold;
    box-shadow: 0 0 10px rgba(255, 0, 0, 0.2);
}

.loader {
    width: 50px;
    height: 50px;
    border: 5px solid #f3f3f3;
    border-top: 5px solid #009879;
    border-radius: 50%;
    animation: spin 1s linear infinite;
    margin: 20px auto;
}

@keyframes spin {
    0% {
        transform: rotate(0deg);
    }
    100% {
        transform: rotate(360deg);
    }
}

