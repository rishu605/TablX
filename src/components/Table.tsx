import React, { useEffect, useState } from "react"
import { getProjectDetailsFromApi } from "../api/api"
import "./Table.css"

interface TableProps {}

interface ProjectDetails {
    "s.no": number;
    "amt.pledged": number;
    "blurb": string;
    "by": string;
    "country": string;
    "currency": string;
    "end.time": string;
    "location": string;
    "percentage.funded": number;
    "num.backers": string;
    "state": string;
    "title": string;
    "type": string;
    "url": string;
}

const Table: React.FC<TableProps> = () => {
    const [projectDetails, setProjectDetails] = useState<ProjectDetails[] | []>([])

    useEffect(() => {
        const fetchProjectDetails = async () => {
            try {
                const apiData = await getProjectDetailsFromApi()
                setProjectDetails(apiData)
            } catch (error) {
                console.error("Failed to fetch project details:", error)
            }
        }

        fetchProjectDetails()
    }, [])

    const renderProjectDetails = projectDetails.map(item => {
        return (
            <tr key={item["s.no"]}>
                <td>{item["s.no"]}</td>
                <td>{item["percentage.funded"]}</td>
                <td>{item["amt.pledged"]}</td>
            </tr>
        )
    })

    return (
        <div className="table-container">
            <h2>Project Details</h2>
            <table className="styled-table">
                <thead>
                    <tr>
                        <th>S.No.</th>
                        <th>Percentage funded</th>
                        <th>Amount pledged</th>
                    </tr>
                </thead>
                <tbody>
                    {renderProjectDetails}
                </tbody>
            </table>
        </div>
    )
}

export default Table