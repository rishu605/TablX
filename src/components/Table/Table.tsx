import React, { useEffect, useState, Fragment } from "react"
import { getProjectDetailsFromApi } from "../../api/api"
import "./Table.css"

interface TableProps {}

interface ProjectDetails {
    "s.no": number;
    "amt.pledged": number;
    "percentage.funded": number;
}

const Table: React.FC<TableProps> = () => {
    const [projectDetails, setProjectDetails] = useState<ProjectDetails[]>([])
    const [currentPage, setCurrentPage] = useState(1)
    const [currentPageGroup, setCurrentPageGroup] = useState(0)
    const recordsPerPage = 5
    const pagesPerGroup = 5
    const [err, setErr] = useState<string>("")

    useEffect(() => {
        const fetchProjectDetails = async () => {
            try {
                const apiData = await getProjectDetailsFromApi()
                setProjectDetails(apiData)
            } catch (error) {
                setErr("Failed to fetch project details")
                console.error("Failed to fetch project details:", error)
            }
        }

        fetchProjectDetails()
    }, [])

    const handlePageChange = (pageNumber: number) => {
        setCurrentPage(pageNumber)
        setCurrentPageGroup(Math.floor((pageNumber - 1) / pagesPerGroup))
    }

    const handleNextPageGroup = () => {
        const nextPage = currentPage + 1
        setCurrentPage(nextPage)
        if (nextPage > (currentPageGroup + 1) * pagesPerGroup) {
            setCurrentPageGroup(currentPageGroup + 1)
        }
    }

    const handlePreviousPageGroup = () => {
        const prevPage = currentPage - 1
        setCurrentPage(prevPage)
        if (prevPage <= currentPageGroup * pagesPerGroup) {
            setCurrentPageGroup(currentPageGroup - 1)
        }
    }

    const renderProjectDetails = () => {
        const currentPageDetails = projectDetails.slice((currentPage - 1) * recordsPerPage, currentPage * recordsPerPage)
        return currentPageDetails
            .map(item => (
                <tr role="row" key={item["s.no"]}>
                    <td role="cell" className="col-width-small">{item["s.no"]}</td>
                    <td role="cell">{item["percentage.funded"]}</td>
                    <td role="cell">{item["amt.pledged"]}</td>
                </tr>
            ))
    }

    const renderTable = () => (
        <table role="table" className="styled-table" aria-label="Project details">
            <thead>
                <tr role="row">
                    <th scope="col" className="col-width-small">S.No.</th>
                    <th scope="col">Percentage funded</th>
                    <th scope="col">Amount pledged</th>
                </tr>
            </thead>
            <tbody>
                {renderProjectDetails()}
            </tbody>
        </table>
    )

    const renderPagination = () => {
        const totalPages = Math.ceil(projectDetails.length / recordsPerPage)
        const startPage = currentPageGroup * pagesPerGroup + 1
        const endPage = Math.min(startPage + pagesPerGroup - 1, totalPages)

        return (
            <div className="pagination">
                <button role="button" aria-label="Previous Page/Group" disabled={currentPage === 1} onClick={handlePreviousPageGroup}>{"<"}</button>
                {Array.from({ length: endPage - startPage + 1 }, (_, index) => (
                    <button
                        key={startPage + index}
                        onClick={() => handlePageChange(startPage + index)}
                        className={currentPage === startPage + index ? "active" : ""}
                    >
                        {startPage + index}
                    </button>
                ))}
                <button aria-label="Next Page/Group" disabled={currentPage === totalPages} onClick={handleNextPageGroup}>{">"}</button>
            </div>
        )
    }

    return (
        <main className="table-container">
            <h2 role="heading" aria-level={2}>Project Details</h2>
            {err ? (
                <div role="alert" className="error-message">
                    {err}
                </div>
            ): (
                <Fragment>
                    {renderTable()}
                    {renderPagination()}
                </Fragment>
            )}
        </main>
    )
}

export default Table