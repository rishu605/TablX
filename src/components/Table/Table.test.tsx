import { render, screen, fireEvent, waitFor, within } from '@testing-library/react';
import Table from './Table';
import { getProjectDetailsFromApi } from '../../api/api';

jest.mock('../../api/api', () => ({
  getProjectDetailsFromApi: jest.fn(),
}));

describe('Table Component', () => {
  beforeEach(() => {
    (getProjectDetailsFromApi as jest.Mock).mockResolvedValue([
      { "s.no": 1, "amt.pledged": 100, "percentage.funded": 10 },
      { "s.no": 2, "amt.pledged": 200, "percentage.funded": 20 },
      { "s.no": 3, "amt.pledged": 300, "percentage.funded": 30 },
      { "s.no": 4, "amt.pledged": 400, "percentage.funded": 40 },
      { "s.no": 5, "amt.pledged": 500, "percentage.funded": 50 },
      { "s.no": 6, "amt.pledged": 600, "percentage.funded": 60 },
    ]);
  });

  it('matches the snapshot', async () => {
    const { container } = render(<Table />);

    // Wait for async data to be loaded
    await new Promise((resolve) => setTimeout(resolve, 0)); // Wait for the microtask queue to clear

    // Snapshot the rendered output
    expect(container).toMatchSnapshot();
  });

  it('renders the table with data', async () => {
    render(<Table />);

    await waitFor(() => screen.getByRole('heading', { name: 'Project Details' }));

    // Verify table rows
    const table = screen.getByRole('table');
    const rows = within(table).getAllByRole('row'); // Includes header and data rows
    expect(rows).toHaveLength(6); // 1 header + 5 data rows

    // Verify specific data in the first row
    const firstRowCells = within(rows[1]).getAllByRole('cell');
    expect(firstRowCells[0]).toHaveTextContent('1'); // S.No
    expect(firstRowCells[1]).toHaveTextContent('10'); // Percentage funded
    expect(firstRowCells[2]).toHaveTextContent('100'); // Amount pledged
  });

  it('renders pagination controls', async () => {
    render(<Table />);

    await waitFor(() => screen.getByRole('heading', { name: 'Project Details' }));

    // Verify pagination container
    const buttons = screen.getAllByRole('button');
    expect(buttons).toHaveLength(4);
  });

  it('handles pagination correctly', async () => {
    render(<Table />);

    await waitFor(() => screen.getByRole('heading', { name: 'Project Details' }));

    // Verify first page rows
    const table = screen.getByRole('table');
    const rows = within(table).getAllByRole('row');
    expect(rows).toHaveLength(6);

    // Verify specific data on the first page
    const firstRowCells = within(rows[1]).getAllByRole('cell');
    expect(firstRowCells[0]).toHaveTextContent('1');

    // Click "Next" button
    const nextButton = screen.getByRole('button', { name: '>' });
    fireEvent.click(nextButton);

    // Wait for the second page to load
    await waitFor(() => {
      const secondPageRows = within(table).getAllByRole('row');
      expect(secondPageRows).toHaveLength(2); // Header + remaining 1 row on page 2
    });

    // Verify data on the second page
    const secondPageFirstRow = within(rows[1]).getAllByRole('cell');
    expect(secondPageFirstRow[0]).not.toHaveTextContent('6');
  });
});
