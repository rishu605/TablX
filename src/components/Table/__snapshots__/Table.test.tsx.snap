// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component matches the snapshot 1`] = `
<div>
  <div
    class="table-container"
  >
    <h2
      role="heading"
    >
      Project Details
    </h2>
    <table
      class="styled-table"
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            class="col-width-small"
          >
            S.No.
          </th>
          <th>
            Percentage funded
          </th>
          <th>
            Amount pledged
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <div
      class="pagination"
    >
      <button
        disabled=""
        role="button"
      >
        &lt;
      </button>
      <button>
        &gt;
      </button>
    </div>
  </div>
</div>
`;
